Thanks. The CI builds for #2520 fail, because `QRgb` (a typedef for `unsigned int`) cannot be used to construct a mixxx::RgbColor. Would you add a function for that as well or should I add a `static_cast<quint32>` instead?

```
src/preferences/configobject.cpp: In member function ‘ResultType ConfigObject<ValueType>::getValue(const ConfigKey&, const ResultType&) const [with ResultType = std::experimental::fundamentals_v1::optional<mixxx::RgbColor>; ValueType = ConfigValue]’:
src/preferences/configobject.cpp:352:63: error: no matching function for call to ‘std::experimental::fundamentals_v1::optional<mixxx::RgbColor>::optional(QRgb)’
     auto color = mixxx::RgbColor::optional_t(parsedColor.rgb());
                                                               ^
In file included from src/util/optional.h:10:0,
                 from src/util/color/rgbcolor.h:9,
                 from src/preferences/configobject.cpp:10:
/usr/include/c++/5/experimental/optional:477:11: note: candidate: constexpr std::experimental::fundamentals_v1::optional<mixxx::RgbColor>::optional()
     class optional
           ^
/usr/include/c++/5/experimental/optional:477:11: note:   candidate expects 0 arguments, 1 provided
/usr/include/c++/5/experimental/optional:503:20: note: candidate: constexpr std::experimental::fundamentals_v1::optional<mixxx::RgbColor>::optional(std::experimental::fundamentals_v1::nullopt_t)
       using _Base::_Base;
                    ^
/usr/include/c++/5/experimental/optional:503:20: note:   no known conversion for argument 1 from ‘QRgb {aka unsigned int}’ to ‘std::experimental::fundamentals_v1::nullopt_t’
/usr/include/c++/5/experimental/optional:503:20: note: candidate: constexpr std::experimental::fundamentals_v1::optional<mixxx::RgbColor>::optional(const mixxx::RgbColor&)
/usr/include/c++/5/experimental/optional:503:20: note:   no known conversion for argument 1 from ‘QRgb {aka unsigned int}’ to ‘const mixxx::RgbColor&’
/usr/include/c++/5/experimental/optional:503:20: note: candidate: constexpr std::experimental::fundamentals_v1::optional<mixxx::RgbColor>::optional(mixxx::RgbColor&&)
/usr/include/c++/5/experimental/optional:503:20: note:   no known conversion for argument 1 from ‘QRgb {aka unsigned int}’ to ‘mixxx::RgbColor&&’
/usr/include/c++/5/experimental/optional:503:20: note: candidate: template<class ... _Args> constexpr std::experimental::fundamentals_v1::optional<mixxx::RgbColor>::optional(std::experimental::fundamentals_v1::in_place_t, _Args&& ...)
/usr/include/c++/5/experimental/optional:503:20: note:   template argument deduction/substitution failed:
src/preferences/configobject.cpp:352:61: note:   cannot convert ‘parsedColor.QColor::rgb()’ (type ‘QRgb {aka unsigned int}’) to type ‘std::experimental::fundamentals_v1::in_place_t’
     auto color = mixxx::RgbColor::optional_t(parsedColor.rgb());
```

--- a/src/control/convert.h
+++ b/src/control/convert.h
@@ -1,0 +1,45 @@
+#pragma once
+
+#include "util/color/rgbcolor.h"
+#include "util/assert.h"
+
+namespace mixxx {
+
+namespace control {
+
+typedef double value_t;
+
+constexpr value_t kInvalidRgbColor = -1.0;
+
+inline value_t valueFromRgbColor(RgbColor color) {
+    return static_cast<RgbColor::code_t>(color);
+}
+
+inline value_t valueFromRgbColor(RgbColor::optional_t color) {
+    if (color) {
+        return valueFromRgbColor(*color);
+    } else {
+        return kInvalidRgbColor;
+    }
+}
+
+inline RgbColor::optional_t valueToRgbColor(value_t value) {
+    if (value >= 0) {
+        const auto code = static_cast<RgbColor::code_t>(value);
+        // If value is out of range then unused bits will be
+        // discarded when converting into an RgbColor. Nevertheless
+        // check with a debug assertion that the given value
+        // is in range as expected. Out of range values indicate
+        // programming errors in other components.
+        DEBUG_ASSERT(RgbColor::isValidCode(code));
+        return RgbColor::optional(code);
+    } else {
+        // < 0 or NaN (if non-signalling)
+        DEBUG_ASSERT(value == kInvalidRgbColor);
+        return RgbColor::nullopt();
+    }
+}
+
+} // namespace control
+
+} // namespace mixxx

--- a/src/library/basesqltablemodel.cpp
+++ b/src/library/basesqltablemodel.cpp
@@ -725,7 +725,7 @@ QVariant BaseSqlTableModel::data(const QModelIndex& index, int role) const {
     switch (role) {
         case Qt::ToolTipRole:
             if (column == fieldIndex(ColumnCache::COLUMN_LIBRARYTABLE_COLOR)) {
-                value = toQString(mixxx::RgbColor::optional(value));
+                value = mixxx::RgbColor::toQString(mixxx::RgbColor::fromQVariant(value));
             }
             M_FALLTHROUGH_INTENDED;
         case Qt::DisplayRole:

--- a/src/library/basetrackcache.cpp
+++ b/src/library/basetrackcache.cpp
@@ -416,7 +416,7 @@ void BaseTrackCache::getTrackValueForColumn(TrackPointer pTrack,
     } else if (fieldIndex(ColumnCache::COLUMN_LIBRARYTABLE_BPM_LOCK) == column) {
         trackValue.setValue(pTrack->isBpmLocked());
     } else if (fieldIndex(ColumnCache::COLUMN_LIBRARYTABLE_COLOR) == column) {
-        trackValue.setValue(toQVariant(pTrack->getColor()));
+        trackValue.setValue(mixxx::RgbColor::toQVariant(pTrack->getColor()));
     } else if (fieldIndex(ColumnCache::COLUMN_LIBRARYTABLE_COVERART_LOCATION) == column) {
         trackValue.setValue(pTrack->getCoverInfo().coverLocation);
     } else if (fieldIndex(ColumnCache::COLUMN_LIBRARYTABLE_COVERART_HASH) == column ||

--- a/src/library/colordelegate.cpp
+++ b/src/library/colordelegate.cpp
@@ -13,7 +13,7 @@ ColorDelegate::ColorDelegate(QTableView* pTableView)
 }
 
 void ColorDelegate::paintItem(QPainter* painter, const QStyleOptionViewItem& option, const QModelIndex& index) const {
-    const auto color = mixxx::RgbColor::optional(index.data());
+    const auto color = mixxx::RgbColor::fromQVariant(index.data());
 
     if (!color) {
         // Filter out track color that is hidden
@@ -23,7 +23,7 @@ void ColorDelegate::paintItem(QPainter* painter, const QStyleOptionViewItem& opt
         return;
     }
 
-    painter->fillRect(option.rect, toQColor(color));
+    painter->fillRect(option.rect, mixxx::RgbColor::toQColor(color));
 
     // Paint transparent highlight if row is selected
     if (option.state & QStyle::State_Selected) {

--- a/src/library/dao/trackdao.cpp
+++ b/src/library/dao/trackdao.cpp
@@ -463,7 +463,7 @@ namespace {
         pTrackLibraryQuery->bindValue(":tracknumber", track.getTrackNumber());
         pTrackLibraryQuery->bindValue(":tracktotal", track.getTrackTotal());
         pTrackLibraryQuery->bindValue(":filetype", track.getType());
-        pTrackLibraryQuery->bindValue(":color", toQVariant(track.getColor()));
+        pTrackLibraryQuery->bindValue(":color", mixxx::RgbColor::toQVariant(track.getColor()));
         pTrackLibraryQuery->bindValue(":comment", track.getComment());
         pTrackLibraryQuery->bindValue(":url", track.getURL());
         pTrackLibraryQuery->bindValue(":duration", track.getDuration());
@@ -985,7 +985,7 @@ bool setTrackTotal(const QSqlRecord& record, const int column,
 
 bool setTrackColor(const QSqlRecord& record, const int column,
                    TrackPointer pTrack) {
-    pTrack->setColor(mixxx::RgbColor::optional(record.value(column)));
+    pTrack->setColor(mixxx::RgbColor::fromQVariant(record.value(column)));
     return false;
 }
 

--- a/src/test/rgbcolor_test.cpp
+++ b/src/test/rgbcolor_test.cpp
@@ -7,73 +7,78 @@
 namespace mixxx {
 
 TEST(RgbColorTest, fromInvalidQColor) {
-    EXPECT_FALSE(RgbColor::optional(QColor()));
-    EXPECT_EQ(RgbColor::nullopt(), RgbColor::optional(QColor()));
+    EXPECT_FALSE(RgbColor::fromQColor(QColor()));
+    EXPECT_EQ(RgbColor::nullopt(), RgbColor::fromQColor(QColor()));
 }
 
 TEST(RgbColorTest, fromQColorWithoutAlpha) {
-    EXPECT_TRUE(RgbColor::optional(QColor::fromRgb(0x000000)));
-    EXPECT_EQ(0x000000, *RgbColor::optional(QColor::fromRgb(0x000000)));
-    EXPECT_TRUE(RgbColor::optional(QColor::fromRgb(0xFF0000)));
-    EXPECT_EQ(RgbColor::optional(0xFF0000), RgbColor::optional(QColor::fromRgb(0xFF0000)));
-    EXPECT_TRUE(RgbColor::optional(QColor::fromRgb(0x00FF00)));
-    EXPECT_EQ(RgbColor::optional(0x00FF00), RgbColor::optional(QColor::fromRgb(0x00FF00)));
-    EXPECT_TRUE(RgbColor::optional(QColor::fromRgb(0x0000FF)));
-    EXPECT_EQ(RgbColor::optional(0x0000FF), RgbColor::optional(QColor::fromRgb(0x0000FF)));
-    EXPECT_TRUE(RgbColor::optional(QColor::fromRgb(0xFFFFFF)));
-    EXPECT_EQ(RgbColor::optional(0xFFFFFF), RgbColor::optional(QColor::fromRgb(0xFFFFFF)));
+    EXPECT_TRUE(RgbColor::fromQColor(QColor::fromRgb(0x000000)));
+    EXPECT_EQ(RgbColor::optional(0x000000), RgbColor::fromQColor(QColor::fromRgb(0x000000)));
+    EXPECT_TRUE(RgbColor::fromQColor(QColor::fromRgb(0xFF0000)));
+    EXPECT_EQ(RgbColor::optional(0xFF0000), RgbColor::fromQColor(QColor::fromRgb(0xFF0000)));
+    EXPECT_TRUE(RgbColor::fromQColor(QColor::fromRgb(0x00FF00)));
+    EXPECT_EQ(RgbColor::optional(0x00FF00), RgbColor::fromQColor(QColor::fromRgb(0x00FF00)));
+    EXPECT_TRUE(RgbColor::fromQColor(QColor::fromRgb(0x0000FF)));
+    EXPECT_EQ(RgbColor::optional(0x0000FF), RgbColor::fromQColor(QColor::fromRgb(0x0000FF)));
+    EXPECT_TRUE(RgbColor::fromQColor(QColor::fromRgb(0xFFFFFF)));
+    EXPECT_EQ(RgbColor::optional(0xFFFFFF), RgbColor::fromQColor(QColor::fromRgb(0xFFFFFF)));
 }
 
 TEST(RgbColorTest, fromQColorWithAlpha) {
-    EXPECT_TRUE(RgbColor::optional(QColor::fromRgba(0xAA000000)));
-    EXPECT_EQ(RgbColor::optional(0x000000), RgbColor::optional(QColor::fromRgba(0xAA000000)));
-    EXPECT_TRUE(RgbColor::optional(QColor::fromRgba(0xAAFF0000)));
-    EXPECT_EQ(RgbColor::optional(0xFF0000), RgbColor::optional(QColor::fromRgba(0xAAFF0000)));
-    EXPECT_TRUE(RgbColor::optional(QColor::fromRgba(0xAA00FF00)));
-    EXPECT_EQ(RgbColor::optional(0x00FF00), RgbColor::optional(QColor::fromRgba(0xAA00FF00)));
-    EXPECT_TRUE(RgbColor::optional(QColor::fromRgba(0xAA0000FF)));
-    EXPECT_EQ(RgbColor::optional(0x0000FF), RgbColor::optional(QColor::fromRgba(0xAA0000FF)));
-    EXPECT_TRUE(RgbColor::optional(QColor::fromRgba(0xAAFFFFFF)));
-    EXPECT_EQ(RgbColor::optional(0xFFFFFF), RgbColor::optional(QColor::fromRgba(0xAAFFFFFF)));
+    EXPECT_TRUE(RgbColor::fromQColor(QColor::fromRgba(0xAA000000)));
+    EXPECT_EQ(RgbColor::optional(0x000000), RgbColor::fromQColor(QColor::fromRgba(0xAA000000)));
+    EXPECT_TRUE(RgbColor::fromQColor(QColor::fromRgba(0xAAFF0000)));
+    EXPECT_EQ(RgbColor::optional(0xFF0000), RgbColor::fromQColor(QColor::fromRgba(0xAAFF0000)));
+    EXPECT_TRUE(RgbColor::fromQColor(QColor::fromRgba(0xAA00FF00)));
+    EXPECT_EQ(RgbColor::optional(0x00FF00), RgbColor::fromQColor(QColor::fromRgba(0xAA00FF00)));
+    EXPECT_TRUE(RgbColor::fromQColor(QColor::fromRgba(0xAA0000FF)));
+    EXPECT_EQ(RgbColor::optional(0x0000FF), RgbColor::fromQColor(QColor::fromRgba(0xAA0000FF)));
+    EXPECT_TRUE(RgbColor::fromQColor(QColor::fromRgba(0xAAFFFFFF)));
+    EXPECT_EQ(RgbColor::optional(0xFFFFFF), RgbColor::fromQColor(QColor::fromRgba(0xAAFFFFFF)));
 }
 
 TEST(RgbColorTest, toQColor) {
-    EXPECT_EQ(QColor::fromRgb(0x123456), toQColor(RgbColor(0x123456)));
+    EXPECT_EQ(QColor::fromRgb(0x123456),
+            RgbColor::toQColor(RgbColor(0x123456)));
 }
 
 TEST(RgbColorTest, toQColorOptional) {
-    EXPECT_EQ(QColor(), toQColor(RgbColor::nullopt()));
+    EXPECT_EQ(QColor(),
+            RgbColor::toQColor(RgbColor::nullopt()));
     EXPECT_EQ(QColor::fromRgba(0xAABBCCDD),
-            toQColor(RgbColor::nullopt(), QColor::fromRgba(0xAABBCCDD)));
+            RgbColor::toQColor(RgbColor::nullopt(), QColor::fromRgba(0xAABBCCDD)));
     EXPECT_EQ(QColor::fromRgb(0x123456),
-            toQColor(RgbColor::optional(QColor::fromRgba(0xAA123456))));
+            RgbColor::toQColor(RgbColor::fromQColor(QColor::fromRgba(0xAA123456))));
     EXPECT_EQ(QColor::fromRgb(0x123456),
-            toQColor(RgbColor::optional(QColor::fromRgba(0xAA123456)), QColor::fromRgba(0xAABBCCDD)));
+            RgbColor::toQColor(RgbColor::fromQColor(QColor::fromRgba(0xAA123456)), QColor::fromRgba(0xAABBCCDD)));
 }
 
 TEST(RgbColorTest, toQVariant) {
-    EXPECT_EQ(QVariant(0x123456), toQVariant(RgbColor(0x123456)));
+    EXPECT_EQ(QVariant(0x123456),
+            RgbColor::toQVariant(RgbColor(0x123456)));
 }
 
 TEST(RgbColorTest, toQVariantOptional) {
-    EXPECT_EQ(QVariant(), toQVariant(RgbColor::nullopt()));
+    EXPECT_EQ(QVariant(),
+            RgbColor::toQVariant(RgbColor::nullopt()));
     EXPECT_EQ(QVariant(0x123456),
-            toQVariant(RgbColor::optional(QColor::fromRgba(0xAA123456))));
+            RgbColor::toQVariant(RgbColor::fromQColor(QColor::fromRgba(0xAA123456))));
 }
 
 TEST(RgbColorTest, toQString) {
-    EXPECT_EQ(QString("#123456"), toQString(RgbColor(0x123456)));
+    EXPECT_EQ(QString("#123456"),
+            RgbColor::toQString(RgbColor(0x123456)));
 }
 
 TEST(RgbColorTest, toQStringOptional) {
     EXPECT_EQ(QString(),
-            toQString(RgbColor::nullopt()));
+            RgbColor::toQString(RgbColor::nullopt()));
     EXPECT_EQ(QString("None"),
-            toQString(RgbColor::nullopt(), QStringLiteral("None")));
+            RgbColor::toQString(RgbColor::nullopt(), QStringLiteral("None")));
     EXPECT_EQ(QString("#123456"),
-            toQString(RgbColor::optional(QColor::fromRgba(0xAA123456))));
+            RgbColor::toQString(RgbColor::fromQColor(QColor::fromRgba(0xAA123456))));
     EXPECT_EQ(QString("#123456"),
-            toQString(RgbColor::optional(QColor::fromRgba(0xAA123456)), QStringLiteral("None")));
+            RgbColor::toQString(RgbColor::fromQColor(QColor::fromRgba(0xAA123456)), QStringLiteral("None")));
 }
 
 } // namespace mixxx

--- a/src/test/rgbcolor_test.cpp
+++ b/src/test/rgbcolor_test.cpp
@@ -3,6 +3,7 @@
 #include <gtest/gtest.h>
 
 #include "test/mixxxtest.h"
+#include "control/convert.h"
 
 namespace mixxx {
 
@@ -79,6 +80,28 @@ TEST(RgbColorTest, toQStringOptional) {
             RgbColor::toQString(RgbColor::fromQColor(QColor::fromRgba(0xAA123456))));
     EXPECT_EQ(QString("#123456"),
             RgbColor::toQString(RgbColor::fromQColor(QColor::fromRgba(0xAA123456)), QStringLiteral("None")));
+}
+
+TEST(RgbColorTest, fromControlValue) {
+    EXPECT_EQ(RgbColor::nullopt(),
+            control::valueToRgbColor(control::kInvalidRgbColor));
+    EXPECT_EQ(RgbColor::optional(0),
+            control::valueToRgbColor(0));
+    EXPECT_EQ(RgbColor::optional(0xFEDCBA),
+            control::valueToRgbColor(0xFEDCBA));
+}
+
+TEST(RgbColorTest, toControlValue) {
+    EXPECT_EQ(control::kInvalidRgbColor,
+            control::valueFromRgbColor(RgbColor::nullopt()));
+    EXPECT_EQ(0.0,
+            control::valueFromRgbColor(RgbColor(0)));
+    EXPECT_EQ(0.0,
+            control::valueFromRgbColor(RgbColor::optional(0)));
+    EXPECT_EQ(16702650.0,
+            control::valueFromRgbColor(RgbColor::optional(0xFEDCBA)));
+    EXPECT_EQ(16702650.0,
+            control::valueFromRgbColor(RgbColor::optional(0xFEDCBA)));
 }
 
 } // namespace mixxx

--- a/src/util/color/rgbcolor.h
+++ b/src/util/color/rgbcolor.h
@@ -20,7 +20,7 @@ class RgbColor {
     // without an an alpha channel.
     // We are using a separate typedef, because QRgb implicitly
     // includes an alpha channel whereas this type does not!
-    typedef quint32 code_t;
+    typedef QRgb code_t;
 
     static constexpr code_t validateCode(code_t code) {
         return code & kRgbCodeMask;

--- a/src/util/color/rgbcolor.h
+++ b/src/util/color/rgbcolor.h
@@ -36,10 +36,6 @@ class RgbColor {
     explicit constexpr RgbColor(code_t code)
             : m_code(validateCode(code)) {
     }
-    // Explicit conversion from QColor.
-    RgbColor(QColor anyColor, code_t codeIfInvalid)
-            : m_code(anyColorToCode(anyColor, codeIfInvalid)) {
-    }
 
     // Implicit conversion to a color code.
     constexpr operator code_t() const {
@@ -64,43 +60,162 @@ class RgbColor {
     static constexpr optional_t optional(RgbColor color) {
         return std::make_optional(color);
     }
-    static optional_t optional(
-            QColor color,
-            optional_t defaultColor = nullopt()) {
-        if (color.isValid()) {
-            return optional(validateCode(color.rgb()));
+
+    ///////////////////////////////////////////////////////////////////
+    // Conversion functions from/to Qt types
+    ///////////////////////////////////////////////////////////////////
+
+    // Explicit conversion of both non-optional and optional
+    // RgbColor values to/from QColor as overloaded free functions.
+
+    static QColor toQColor(RgbColor color) {
+        return QColor::fromRgb(color);
+    }
+
+    static QColor toQColor(
+            RgbColor::optional_t optional,
+            const QColor& defaultColor = QColor()) {
+        if (optional) {
+            return toQColor(*optional);
         } else {
             return defaultColor;
         }
     }
-    static optional_t optional(
+
+    static RgbColor::optional_t fromQColor(
+            const QColor& color,
+            RgbColor::optional_t defaultColor = RgbColor::nullopt()) {
+        if (color.isValid()) {
+            return RgbColor::optional(
+                    RgbColor::validateCode(color.rgb()));
+        } else {
+            return defaultColor;
+        }
+    }
+
+    // Explicit conversion of both non-optional and optional
+    // RgbColor values to/from QString in the format #RRGGBB,
+    // e.g. for tool tips or settings.
+
+    static QString toQString(
+            mixxx::RgbColor color) {
+        return toQColor(color).name(QColor::HexRgb);
+    }
+
+    static QString toQString(
+            mixxx::RgbColor::optional_t color,
+            const QString& defaultString = QString()) {
+        if (color) {
+            return toQString(*color);
+        } else {
+            return defaultString;
+        }
+    }
+
+    static RgbColor::optional_t fromQString(
+            const QString& hexCode,
+            RgbColor::optional_t defaultColor = RgbColor::nullopt()) {
+        const auto color = QColor(hexCode);
+        if (color.isValid()) {
+            return fromQColor(color);
+        } else {
+            return defaultColor;
+        }
+    }
+
+    // Explicit conversion of both non-optional and optional
+    // RgbColor values to QVariant as overloaded free functions.
+    //
+    // The default versions encode the internal color code as
+    // an unsigned integer. The `Color` and `String` variants
+    // use the corresponding mapping to QColor and QString
+    // respectively.
+
+    static QVariant toQVariant(
+            RgbColor color) {
+        return QVariant(static_cast<code_t>(color));
+    }
+
+    static QVariant toQVariantColor(
+            RgbColor color) {
+        return QVariant(toQColor(color));
+    }
+
+    static QVariant toQVariantString(
+            RgbColor color) {
+        return QVariant(toQString(color));
+    }
+
+    static QVariant toQVariant(
+            optional_t optional,
+            const QVariant& defaultVariant = QVariant()) {
+        if (optional) {
+            return toQVariant(*optional);
+        } else {
+            return defaultVariant;
+        }
+    }
+
+    static QVariant toQVariantColor(
+            optional_t optional,
+            const QVariant& defaultVariant = QVariant()) {
+        if (optional) {
+            return toQVariantColor(*optional);
+        } else {
+            return defaultVariant;
+        }
+    }
+
+    static QVariant toQVariantString(
+            optional_t optional,
+            const QVariant& defaultVariant = QVariant()) {
+        if (optional) {
+            return toQVariantString(*optional);
+        } else {
+            return defaultVariant;
+        }
+    }
+
+    static optional_t fromQVariant(
             const QVariant& varCode,
             optional_t defaultColor = nullopt()) {
         if (varCode.isNull()) {
             return defaultColor;
-        } else {
-            DEBUG_ASSERT(varCode.canConvert(QMetaType::UInt));
-            bool ok = false;
-            const auto code = varCode.toUInt(&ok);
-            VERIFY_OR_DEBUG_ASSERT(ok) {
-                return defaultColor;
-            }
-            return optional(static_cast<code_t>(code));
         }
+        DEBUG_ASSERT(varCode.canConvert(QMetaType::UInt));
+        bool ok = false;
+        const auto code = varCode.toUInt(&ok);
+        VERIFY_OR_DEBUG_ASSERT(ok) {
+            return defaultColor;
+        }
+        return optional(static_cast<code_t>(code));
+    }
+
+    static optional_t fromQVariantColor(
+            const QVariant& varColor,
+            optional_t defaultColor = nullopt()) {
+        if (varColor.isNull()) {
+            return defaultColor;
+        }
+        DEBUG_ASSERT(varColor.canConvert(QMetaType::QColor));
+        const auto value = varColor.value<QColor>();
+        return fromQColor(value);
+    }
+
+    static optional_t fromQVariantString(
+            const QVariant& varString,
+            optional_t defaultColor = nullopt()) {
+        if (varString.isNull()) {
+            return defaultColor;
+        }
+        DEBUG_ASSERT(varString.canConvert(QMetaType::QString));
+        const auto value = varString.value<QString>();
+        return fromQString(value);
     }
 
   protected:
     // Bitmask of valid codes = 0x00RRGGBB
     static constexpr code_t kRgbCodeMask = 0x00FFFFFF;
-
-    static code_t anyColorToCode(QColor anyColor, code_t codeIfInvalid) {
-        if (anyColor.isValid()) {
-            // Strip alpha channel bits!
-            return validateCode(anyColor.rgb());
-        } else {
-            return codeIfInvalid;
-        }
-    }
 
     code_t m_code;
 };
@@ -109,69 +224,14 @@ inline bool operator!=(RgbColor lhs, RgbColor rhs) {
     return !(lhs == rhs);
 }
 
-// Explicit conversion of both non-optional and optional
-// RgbColor values to QColor as overloaded free functions.
-
-inline QColor toQColor(RgbColor color) {
-    return QColor::fromRgb(color);
-}
-
-inline QColor toQColor(
-        RgbColor::optional_t optional,
-        QColor defaultColor = QColor()) {
-    if (optional) {
-        return toQColor(*optional);
-    } else {
-        return defaultColor;
-    }
-}
-
-// Explicit conversion of both non-optional and optional
-// RgbColor values to QVariant as overloaded free functions.
-
-inline QVariant toQVariant(
-        RgbColor color) {
-    return QVariant(static_cast<RgbColor::code_t>(color));
-}
-
-inline QVariant toQVariant(
-        RgbColor::optional_t optional,
-        const QVariant& defaultVariant = QVariant()) {
-    if (optional) {
-        return toQVariant(*optional);
-    } else {
-        return defaultVariant;
-    }
-}
-
-// Explicit conversion of both non-optional and optional
-// RgbColor values to QString with the format #RRGGBB,
-// e.g. for tool.
-
-inline QString toQString(
-        mixxx::RgbColor color) {
-    return mixxx::toQColor(color).name();
-}
-
-// String representation #RRGGBB, e.g. for tool.
-inline QString toQString(
-        mixxx::RgbColor::optional_t color,
-        const QString& defaultString = QString()) {
-    if (color) {
-        return toQString(*color);
-    } else {
-        return defaultString;
-    }
-}
-
 // Debug output stream operators
 
 inline QDebug operator<<(QDebug dbg, RgbColor color) {
-    return dbg << toQColor(color);
+    return dbg << RgbColor::toQString(color).toLatin1().constData();
 }
 
-inline QDebug operator<<(QDebug dbg, RgbColor::optional_t optional) {
-    return dbg << toQColor(optional);
+inline QDebug operator<<(QDebug dbg, RgbColor::optional_t optionalColor) {
+    return dbg << RgbColor::toQString(optionalColor).toLatin1().constData();
 }
 
 } // namespace mixxx

--- a/src/widget/wtracktableview.cpp
+++ b/src/widget/wtracktableview.cpp
@@ -1093,7 +1093,7 @@ void WTrackTableView::contextMenuEvent(QContextMenuEvent* event) {
             // Get color of first selected track
             int column = trackModel->fieldIndex(LIBRARYTABLE_COLOR);
             QModelIndex index = indices.at(0).sibling(indices.at(0).row(), column);
-            mixxx::RgbColor::optional_t trackColor = mixxx::RgbColor::optional(index.data());
+            auto trackColor = mixxx::RgbColor::fromQVariant(index.data());
 
             // Check if all other selected tracks have the same color
             bool multipleTrackColors = false;
@@ -1101,9 +1101,8 @@ void WTrackTableView::contextMenuEvent(QContextMenuEvent* event) {
                 int row = indices.at(i).row();
                 QModelIndex index = indices.at(i).sibling(row, column);
 
-                mixxx::RgbColor::optional_t otherTrackColor = mixxx::RgbColor::optional(index.data());
-                if (trackColor != otherTrackColor) {
-                    trackColor = std::nullopt;
+                if (trackColor != mixxx::RgbColor::fromQVariant(index.data())) {
+                    trackColor = mixxx::RgbColor::nullopt();
                     multipleTrackColors = true;
                     break;
                 }
@@ -1996,7 +1995,7 @@ void WTrackTableView::slotColorPicked(PredefinedColorPointer pColor) {
     // TODO: This should be done in a thread for large selections
     for (const auto& index : selectedTrackIndices) {
         TrackPointer track = trackModel->getTrack(index);
-        track->setColor(mixxx::RgbColor::optional(pColor->m_defaultRgba));
+        track->setColor(mixxx::RgbColor::fromQColor(pColor->m_defaultRgba));
     }
 
     m_pMenu->hide();

